File: LinkedList/ReverseLLInEvenParts/reverseLL.cpp
#include <iostream>

struct ListNode {
public:
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    ListNode* reverse(ListNode* head){
        ListNode* Next = head;
        ListNode* curr = head;
        ListNode* prev = NULL;
        while(curr){
            Next = curr->next;
            curr->next = prev;
            prev = curr;
            curr = Next;
        }
        return prev;
    }
    ListNode* reversePart(ListNode*  head,int left,int right){
        ListNode* a = NULL;
        ListNode* b = NULL;
        ListNode* c = NULL;
        ListNode* d = NULL;
        ListNode* temp = head;
        int n = 1;
        while(temp){
            if(left-1==n)a = temp;
            else if(left==n)b = temp;
            else if(right==n)c = temp;
            else if(right+1==n)d = temp;
            temp = temp->next;
            n++;
        }
        if(a)a->next = NULL;
        c->next = NULL;
        c = reverse(b);
        if(a)a->next = c;
        b->next = d;
        if(a)return a;
        return c;
    }
    ListNode* reverseEvenLengthGroups(ListNode* head) {
        int gap = 1;
        ListNode* temp = head;
        while(temp && temp->next){
            ListNode* t = temp->next;
            int remLen = 0;
            for(int i=1;i<=gap+1 && t!=NULL;i++){
                t=t->next;
                remLen++;
            }
            if(remLen<gap+1)gap = remLen-1;
            if(gap%2!=0)reversePart(temp,2,gap+2);
            gap++;
            for(int i=1;temp!=NULL && i<=gap;i++){
                temp=temp->next;
            }
        }
        return head;
    }
};

File: LinkedList/ReverseLLInEvenParts/reverseLL.py
from typing import Optional

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
        
class Solution:
    def reverseList(self, head: Optional[ListNode]):
        if head==None or head.next==None: return head
        newHead = self.reverseList(head.next)
        head.next.next = head
        head.next = None
        return newHead

    def reverseBetween(self, head: Optional[ListNode], left: int, right: int) -> Optional[ListNode]:
        if left == right: return head
        n = 1
        temp = head
        a = None
        b = None
        c = None
        d = None
        while temp and n < right+1:
            if n==left-1:
                a = temp
            elif n==left:
                b = temp
            elif n==right:
                c = temp
            n += 1
            temp = temp.next
        d = c.next
        c.next = None
        t = self.reverseList(b)
        if a: a.next = t
        else: head = t
        b.next = d
        return head

    def reverseEvenLengthGroups(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head or not head.next: return head
        gap = 2
        temp = head
        while temp:
            t = temp.next
            remLen = 0
            while remLen < gap and t:
                t=t.next
                remLen+=1
            t = temp.next
            if remLen == 0: break
            if remLen%2 == 0: temp.next = self.reverseBetween(t,1,remLen)
            count = remLen
            while count:
                if temp: 
                    temp = temp.next
                    count -= 1
                else: break
            gap += 1
        return head
